feat(core): restrict deletion of the attribute definition

HejdaJakub opened the pull request:
* We want to restrict the deletion of the attribute definition if this attribute is required for any service.
* We want to restrict it also for the case when this attribute is a source or a destination attribute of any application form item in some application form.

xflord reviewed (commented):
Also please fix the tests that now fail

sarkapalkovicova commented on perun-core/src/main/java/cz/metacentrum/perun/core/api/AttributesManager.java:
```diff
@@ -1733,7 +1733,7 @@ Map<String, Attribute> getEntitylessAttributesWithKeys(PerunSession sess, String
 	 * @throws PrivilegeException     if privileges are not given
 	 */
-	void deleteAttribute(PerunSession sess, AttributeDefinition attributeDefinition) throws PrivilegeException, AttributeNotExistsException;
+	void deleteAttribute(PerunSession sess, AttributeDefinition attributeDefinition) throws PrivilegeException, AttributeNotExistsException, RelationExistsException;
```
Maybe should be `RelationExistsException` also mentioned in `AttributesManagerMethod.java` in the doc for this method?

HejdaJakub commented:
> Also please fix the tests that now fail

I am not sure what the strong/inverse/all dependencies are in the context of these failing tests, so I am not sure how to properly fix it. :disappointed:  But these tests are saying that deleted attribute is required for any service, so maybe it is correct that these test doesn't work and should be marked as `@Ignore` or deleted now?

HejdaJakub commented:
I also tried to improve error messages for better readability (I used to paste long objects into this message):
Example of error for service required attribute:
`Attribute definition with id: 3584 is the required attribute for these services: [AttributesManagerTestService (id: 862), AttributesManagerTestService2 (id: 863)]`

Example of error for app from items relation (first for group form, second for vo form):
`Attribute definition with id: 3589 is source or destination attribute in these application forms: [Application form in Group with id: 9645 (under Vo with id: 3323), Application form in Vo with id: 3323]`

HejdaJakub commented:
I tried to improve the error message for the app form item relations use-case again. Now the message contains also `shortname` of the app form items that are related to the attribute. There is also one new check related to the hidden/disabled dependencies. If e.g. `item1` has a dependency on `item2` with source/destination attribute to delete, after removing source/destination attribute from `item2` there won't be any blocking relation, but we want at least inform the user in the error message, that this `item1` has also some weak relation to the attribute which should be managed (because it can break application form logic). The error message now doesn't handle each use case (source / destination / hidden or disabled dependency) separately, rather just informs generally that items in the message have some relation.


`Attribute definition with id: 3589 has a relation (as a source/destination/dependency) to these application form items: [Application form items: [name, displayName] in the application form in Group with id: 9645 (under Vo with id: 3323), Application form items: [shortname, displayName] in the application form in Group with id: 9646 (under Vo with id: 3323)]`

xflord commented on perun-core/src/main/java/cz/metacentrum/perun/core/impl/AttributesManagerImpl.java:
```diff
@@ -4215,6 +4240,44 @@ public List<String> getAllNamespaces(PerunSession sess) {
+	@Override
+	public List<ApplicationForm> getAppFormsWhereAttributeRelated(PerunSession sess, AttributeDefinition attr) {
+		try {
+			String urn = attr.getNamespace() + ":" + attr.getFriendlyName();
```
just realized there's a method for this (`attr.getName()`)

xflord commented on perun-core/src/main/java/cz/metacentrum/perun/core/impl/AttributesManagerImpl.java:
```diff
@@ -4215,6 +4240,44 @@ public List<String> getAllNamespaces(PerunSession sess) {
+	@Override
+	public List<String> getAppFormItemsForAppFormAndAttribute(PerunSession sess, int appFormId, AttributeDefinition attr) {
+		try {
+			String urn = attr.getNamespace() + ":" + attr.getFriendlyName();
```
same here

xflord commented:
> There is also one new check related to the hidden/disabled dependencies. If e.g. `item1` has a dependency on `item2` with source/destination attribute to delete, after removing source/destination attribute from `item2` there won't be any blocking relation, but we want at least inform the user in the error message, that this `item1` has also some weak relation to the attribute which should be managed (because it can break application form logic). The error message now doesn't handle each use case (source / destination / hidden or disabled dependency) separately, rather just informs generally that items in the message have some relation.

This to me feels a bit confusing. I feel like the exception should only include the blocking relations, any warnings related to removing form items should be displayed WHEN removing form items (maybe we should add a warning when removing form items that other items have a dependency on?)

HejdaJakub commented:
> This to me feels a bit confusing. I feel like the exception should only include the blocking relations, any warnings related to removing form items should be displayed WHEN removing form items (maybe we should add a warning when removing form items that other items have a dependency on?)

It makes sense to me, so I removed this warning and moved it to the front end for better UX. 

Just FYI - at the front end part, I am checking the deletion of an app item and I also warn a user when he/she tries to change the source or destination attribute of an item that is marked as a dependency of another item. This should cover all main use cases related to this issue IMHO.
